### PR TITLE
[PM-10311] Account Management: Create helper methods for checking against verified domains

### PR DESCRIPTION
**Pull request by r-tome**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-10311

## 📔 Objective

The Account Management MVP will introduce managed and unmanaged members. This PR aims to add server methods to identify managed users.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: src/Core/AdminConsole/Services/IOrganizationService.cs
Comment (addisonbeck) [inline src/Core/AdminConsole/Services/IOrganizationService.cs:108]: Lets put this in a query instead of growing `OrganizationService`

File: src/Core/Services/IOrganizationDomainService.cs
Comment (addisonbeck) [inline src/Core/Services/IOrganizationDomainService.cs:10]: Should `OrganizationDomainService` be in the `AdminConsole` folder for code ownership purposes? We can move it without changing the namespace to avoid breaking imports if there are a lot of them, but while we're here it makes sense to move it if this a part of ACs domain.

File: src/Core/Services/Implementations/OrganizationDomainService.cs
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationDomainService.cs:56]:
```suggestion
                    // Update entry on OrganizationDomain table
```
You're already removing whitespace at the end of these comments 🤷‍♂️

File: src/Core/Services/Implementations/OrganizationDomainService.cs
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationDomainService.cs:67]:
```suggestion
                    // Update entry on OrganizationDomain table
```

File: src/Core/Services/Implementations/OrganizationDomainService.cs
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationDomainService.cs:81]:
```suggestion
                // Update entry on OrganizationDomain table
```

File: src/Core/Services/Implementations/OrganizationDomainService.cs
Comment (addisonbeck) [inline src/Core/Services/Implementations/OrganizationDomainService.cs:120]:
```suggestion
            // Delete domains that have not been verified within 7 days
```

**Review by addisonbeck — COMMENTED** [review]
Quick feedback. This wasn't really a comprehensive review - I'll get back to it.

File: src/Core/Services/IOrganizationDomainService.cs
Comment (r-tome) [reply src/Core/Services/IOrganizationDomainService.cs]: I moved the namespace, there were only two references.

File: src/Core/AdminConsole/Services/IOrganizationService.cs
Comment (r-tome) [reply src/Core/AdminConsole/Services/IOrganizationService.cs]: Great idea!

**Review by eliykat — CHANGES_REQUESTED** [review]
I have a few quick comments for things that jumped out at me but will take a closer look tomorrow.

**Review by cyprain-okeke — DISMISSED** [review]
Looks good for Billing

File: src/Core/AdminConsole/Models/Data/Organizations/OrganizationAbility.cs
Comment (eliykat) [inline src/Core/AdminConsole/Models/Data/Organizations/OrganizationAbility.cs:24]: I forgot that plan features are replicated in the `Organization` table and tracked separately for each org. I never really understood why we did that, except to enable Custom plans, which we don't hand out anyway (to my knowledge).

I'll follow up to see what we should do here generally (do we need to add another column? Do we actually need these columns?) but for now I think this approach is OK, even though it's breaking the pattern a little.

My only request is to handle a `null` plan, it's possible (although unlikely) that an org has an invalid plan, and in that case I'd rather default to `false` than throw an exception.

Also, for consistent wording, `UseOrganizationDomains`.

**Review by eliykat — CHANGES_REQUESTED** [review]
Thanks! 1 minor thing I spotted, otherwise I don't want to hold this up any further so happy for @addisonbeck to approve when he's happy.

File: src/Core/AdminConsole/Models/Data/Organizations/OrganizationAbility.cs
Comment (r-tome) [reply src/Core/AdminConsole/Models/Data/Organizations/OrganizationAbility.cs:24]: Good question on Custom plans. I looked up how we were gating Domain Verification on the client side and we just look at the `UseSso` field https://github.com/bitwarden/clients/blob/819c312ce272c4b33efbfdab9618ba090b294597/libs/common/src/admin-console/models/domain/organization.ts#L255

So I'm thinking maybe we don't even need a new field at all, just look at `UseSso`. What do you think?

Comment (eliykat) [reply src/Core/AdminConsole/Models/Data/Organizations/OrganizationAbility.cs]: Domain Verification started out as an SSO feature - it makes the login flow smoother - but these new features are no longer related to SSO. You could use a verified domain to manage account deprovisioning better even without using SSO. So I think it's become its own feature.

But for now, given that we're not ready to introduce new columns over it, let's stick with the current pattern of using `UseSso` for this. I'll create a follow up ticket so we can come back and adjust later if required.

File: src/Core/Services/Implementations/UserService.cs
Comment (eliykat) [inline src/Core/Services/Implementations/UserService.cs:1255]: This also needs updating to 
* use applicationCacheService
* refer to `useSso`
* flatten nesting

Comment (r-tome) [reply src/Core/Services/Implementations/UserService.cs]: I can't use `IApplicationCacheService` here, I need to use `OrganizationRepository.GetByClaimedUserDomainAsync` to get the organization that has a verified domain that matches the User's email domain

Comment (eliykat) [reply src/Core/Services/Implementations/UserService.cs]: You're right! I misread it.

**Review by eliykat — DISMISSED** [review]

**Comment by eliykat** [comment]
(But please get @addisonbeck's final review)

File: src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/GetOrganizationUsersManagementStatusQuery.cs
Comment (addisonbeck) [inline src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/GetOrganizationUsersManagementStatusQuery.cs:26]: Since we've identified that `UseSso` is not very apt here and are going to improve this as tech debt: lets add a comment here indicating that `UseSso` is used to signify that the organization has access to domain verification. Bonus points if it includes a link to a Jira ticket.

**Review by addisonbeck — COMMENTED** [review]
Very small nitpick.

File: src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/GetOrganizationUsersManagementStatusQuery.cs
Comment (eliykat) [reply src/Core/AdminConsole/OrganizationFeatures/OrganizationUsers/GetOrganizationUsersManagementStatusQuery.cs:26]: I said I'd follow this up, so I've created a ticket for it here that you can link to: https://bitwarden.atlassian.net/browse/PM-11622

**Review by addisonbeck — APPROVED** [review]

File: util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql:7]: Would benefit from index:
```
CREATE NONCLUSTERED INDEX [<index_name>]
ON [dbo].[OrganizationDomain] ([OrganizationId],[VerifiedDate])
```

File: util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql:27]: Would benefit from index
```
CREATE NONCLUSTERED INDEX [<index name>]
ON [dbo].[OrganizationDomain] ([VerifiedDate])
INCLUDE ([OrganizationId],[DomainName])
```

**Review by addisonbeck — APPROVED** [review]

File: util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql]: this is default behavior in cloud; WITH (ONLINE = ON); no need to add

File: util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql
Comment (rkac-bw) [inline util/Migrator/DbScripts/2024-09-04_00_UsersManagedByOrg.sql]: this is default behavior in cloud; WITH (ONLINE = ON); no need to add

**Review by rkac-bw — APPROVED** [review]
lgtm

**Review by cyprain-okeke — APPROVED** [review]

**Pull request merged**
